rs/replicated_state/src/canister_state/queues/refunds.rs

```diff
@@ -0,0 +1,88 @@
+use ic_types::messages::Refund;
+use ic_types::{CanisterId, Cycles};
+use ic_validate_eq::ValidateEq;
+use ic_validate_eq_derive::ValidateEq;
+use std::collections::btree_map::Entry::{Occupied, Vacant};
+use std::collections::{BTreeMap, BTreeSet};
+
+#[cfg(test)]
+mod tests;
+
+/// A prioritized pool of refunds to canisters. Used for accumulating outbound
+/// refunds at the subnet level, before routing into streams.
+///
+/// Refunds are ordered by amount (larger amounts first). Ties are broken by
+/// recipient (smaller IDs first).
+#[derive(Debug, Clone, Default, PartialEq, Eq, ValidateEq)]
+pub struct RefundPool {
+    /// Refund priority queue. Holds all refunds, ordered by amount.
+    ///
+    /// Canister IDs break ties, ensuring deterministic ordering.
+    refunds: BTreeSet<Refund>,
+
+    // Refund amounts, by recipient.
+    #[validate_eq(Ignore)]
+    amounts: BTreeMap<CanisterId, Cycles>,
```

Comment:
I considered having both the `BTreeSet` and the `BTreeMap` hold `Arc<Refund>` values instead, so as to make cloning slightly more efficient. But I believe the trade-off is likely not worth it; and the pool will be empty most of the time; and relatively compact anyway.